packages/react-devtools-shared/src/backend/renderer.js

```diff
@@ -1073,6 +1074,48 @@ export function attach(
     return null;
   }
 
+  function areHookInputsEqual(
+    nextDeps: Array<mixed>,
+    prevDeps: Array<mixed> | null,
+  ) {
+    if (prevDeps === null) {
+      return false;
+    }
+
+    for (let i = 0; i < prevDeps.length && i < nextDeps.length; i++) {
+      if (is(nextDeps[i], prevDeps[i])) {
+        continue;
+      }
+      return false;
+    }
+    return true;
+  }
+
+  function isEffect(memoizedState) {
+    return (
+      memoizedState !== null &&
+      typeof memoizedState === 'object' &&
+      memoizedState.hasOwnProperty('tag') &&
+      memoizedState.hasOwnProperty('create') &&
+      memoizedState.hasOwnProperty('destroy') &&
+      memoizedState.hasOwnProperty('deps') &&
+      memoizedState.hasOwnProperty('next')
+    );
+  }
+
+  function didHookChange(prev: any, next: any): boolean {
+    const prevMemoizedState = prev.memoizedState;
+    const nextMemoizedState = next.memoizedState;
+
+    if (isEffect(prevMemoizedState) && isEffect(nextMemoizedState)) {
+      return !areHookInputsEqual(
+        nextMemoizedState.deps,
+        prevMemoizedState.deps,
+      );
+    }
```

Comment:
Not a fan of this new coupling. Especially since this will not work as expected for e.g. `useState({ tag, create, destroy, deps, next });` even though that's unlikely to occur in userland. @bvaughn Any suggestions how this should work now that `hook.memoizedState` always changes even if deps didn't?

Reply:
DevTools probably should be resilient to this happening, however, if you want to kick the can down the road, another solution could be to assign `hook.memoizedState = current.memoizedState` when you bailout. Instead of the result of `pushEffect`. Since for render phase updates, that will have the effect of restoring the original object.

Reply:
If you end up doing this, please leave a TODO here so we remember this is accidentally coupled

Reply:
> But for render phase updates, it will restore the original object.

This was actually my first intuition but I didn't really grasp how `hook` relates to `currentHook`.

But since you seem to be ok with the current approach I'm keeping the initial fix. I added an additional type check to `memoizedState.deps` in case someone has `useState({ deps: 1, tag, create, destroy, next });`

Reply:
Still planning on looking at this question. It's next in my queue.

Reply:
At first I thought this might break the Profiler too (how it determines what rendered and what didn't) but I forgot that uses the special `PerformedWork` tag.